### PR TITLE
 soc: it8xxx2: linker: Move library to RAM to enhance performance

### PR DESCRIPTION
**Pull request by GTLin08**
Three new configuration options are added, including serial, kernel and zephyr libraries, and placed into ILM as RAM code to improve performance.

### CODE REVIEW

File: soc/ite/ec/it8xxx2/Kconfig [hunk 1]
```diff
@@ -183,6 +183,17 @@ config SOC_IT8XXX2_EXCEPTIONS_IN_RAM
 	  Flash. This can significantly improve performance when under I-cache
 	  pressure.
 
+config SOC_IT8XXX2_LIBRARY_TO_RAM
+	bool "Move library to RAM"
```
Comment (keith-zephyr) [inline soc/ite/ec/it8xxx2/Kconfig:187]: This appears to be a helper symbol only.  [Helper symbols ](https://docs.zephyrproject.org/latest/build/kconfig/tips.html#using-select-for-helper-symbols) should not have prompt text, but should provide help text.

File: soc/ite/ec/it8xxx2/linker.ld [hunk 2]
```diff
@@ -176,6 +176,9 @@ SECTIONS
 #ifdef CONFIG_SOC_IT8XXX2_SERIAL_IN_RAM
 			*libdrivers__serial.a:*
 #endif
+#ifdef CONFIG_SOC_IT8XXX2_KERNEL_IN_RAM
+			*libkernel.a:*
```
Comment (keith-zephyr) [inline soc/ite/ec/it8xxx2/linker.ld:180]: Is it possible to include a smaller set of kernel functions than the entire kernel into ILM?  Adding almost 9K to the RAM size is pretty expensive.

File: soc/ite/ec/it8xxx2/linker.ld [hunk 3]
```diff
@@ -179,6 +179,9 @@ SECTIONS
 #ifdef CONFIG_SOC_IT8XXX2_KERNEL_IN_RAM
 			*libkernel.a:*
 #endif
+#ifdef CONFIG_SOC_IT8XXX2_ZEPHYR_IN_RAM
+			*libzephyr.a:*
```
Comment (keith-zephyr) [inline soc/ite/ec/it8xxx2/linker.ld:183]: Same thing here - can you identify a smaller set of library functions to move into ILM?

File: soc/ite/ec/it8xxx2/Kconfig (hunk 1 above)
Comment (GTLin08) [reply soc/ite/ec/it8xxx2/Kconfig]: Done.

File: soc/ite/ec/it8xxx2/linker.ld (hunk 2 above)
Comment (GTLin08) [reply soc/ite/ec/it8xxx2/linker.ld:180]: Here are the reasons for considering the entire libkernel or libzephyr as RAM code:
1. The ILM has increased from 60k in the original IT81xx2 to 256k in the IT82xx2. The main reason for increasing RAM was to account for the potential increase in latency when switching from the legacy RTOS to the Zephyr RTOS. Therefore, in the IT82xx2, there is sufficient space to accommodate the libkernel or libzephyr as RAM code.
2. According to the PD standard specification, `tReceiverResponse`is 15ms. After placing the libkernel and libzephyr as RAM code on the rauru platform, `tReceiverResponse`is reduced from 18ms to 9ms, showing significant improvement and compliance with the specification.
Additionally, the following issues have shown noticeable improvement.
https://partnerissuetracker.corp.google.com/issues/300105688
https://partnerissuetracker.corp.google.com/issues/347824320
3. The libkernel or libzephyr consists of multiple sets. For example, libkernel.a includes `busy_wait.c, device.c, init.c, idle.c, mutex.c, sem.c, system_work_q.c, work.c, thread.c, sched.c, timeout.c, timer.c, poll.c, fatal.c.` Files like `init.c, fatal.c, device.c, etc.,` have minimal impact, but configuring each one individually would result in excessive configurations.

File: soc/ite/ec/it8xxx2/linker.ld (hunk 3 above)
Comment (GTLin08) [reply soc/ite/ec/it8xxx2/linker.ld:183]: As above.

File: soc/ite/ec/it8xxx2/linker.ld (hunk 2 above)
Comment (keith-zephyr) [reply soc/ite/ec/it8xxx2/linker.ld:180]: Acknowledged.

**Review by keith-zephyr — APPROVED** [review]

**Review by Dino-Li — APPROVED** [review]

**Comment by fabiobaltieri** [comment]
@GTLin08 could this be done with ramfunc https://docs.zephyrproject.org/latest/kernel/code-relocation.html? I think that may have some more fined grain control on what goes there, and can be configured in the application rather than on the linker, saves you from having this for every platform.

**Comment by GTLin08** [comment]
> @GTLin08 could this be done with ramfunc https://docs.zephyrproject.org/latest/kernel/code-relocation.html? I think that may have some more fined grain control on what goes there, and can be configured in the application rather than on the linker, saves you from having this for every platform.

Hi @fabiobaltieri ,

Thank you for your suggestion regarding using code relocation. Due to some issues, implementing this PR can effectively resolve them, as mentioned earlier.  Additionally, only platforms using the it82xx2 chip can choose whether to enable this configuration, it is not applicable to every platform. We will prioritize implementing this method for now and then look into your proposed suggestion.

**Pull request merged**